### PR TITLE
ENH: special: add `logiv` and `logive`

### LINKED ISSUE #12607: Adding log(ive) Modified Bessel Function of the first kind
As a follow up to #12244 I had fun adding a function to calculate `log(ive)`.

@ev-br mentioned this is already done in [_hyp0f1.pxd#L51](https://github.com/scipy/scipy/blob/master/scipy/special/_hyp0f1.pxd#L51)
This points to [DLMF 10.41.10](https://dlmf.nist.gov/10.41), I extended _hyp0f1 to k=5 using [A144617](oeis.org/A144617), which reduces max error.

I'd love some advice from @WarrenWeckesser about
* where this code belongs
* if it needs to support complex
* how to make sure we can call it from `stats.ncx2`

---

using `timeit` in colab, `special.ive` takes ~1.29 µs, new pure python code takes `18.1 µs`
This method is O(~1000) operations: 4x `np.log`, 1x `np.sqrt`, 1x `log1p`, ~40 multiplications, and 10 divisions. I expected it to be slightly faster, If I cython we'll have more data.

Error is still large when `v` and `z` are small. Maybe it makes sense to call `np.log(special.ive(v, z))` when v and z are small and only use this when v and z are large.

<details>
<summary>Click to toggle implementation</summary>

```
import scipy.special
import numpy as np

def test_ive_asy(v, z, k=False):
    r"""Asymptotic expansion for I_{v}(z)
    for real $z > 0$ and $v\to +\infty$.
    Based off DLMF 10.41
    """
    # DLMF 10.41.3 calculates I_v(v*z)
    x = z / v

    # DLMF 10.41.8
    p = np.sqrt(1.0 + x*x)

    # DLMF 10.41.7
    eta = p + np.log(x) - scipy.special.log1p(p)

    # e^(vn) / ((2piv)^(1/2)*p^(1/2))
    log_res = v*eta - (np.log(2.0*np.pi*v)/2 + np.log(p)/2)

    # large-v asymptotic correction, DLMF 10.41.10
    pp = 1.0/p
    p2 = pp*pp
    p4 = p2*p2
    p6 = p4*p2
    p8 = p4*p4
    p10 = p6*p4

    v2 = v*v
    v3 = v2*v
    v4 = v2*v2
    v5 = v3*v2

    # u1,u2,u3 from DLMF, higher terms from oeis.org/A144617 and oeis.org/A144618
    u1 = (3.0 - 5.0*p2) * pp / 24.0
    u2 = (81.0 - 462.0*p2 + 385.0*p4) * p2 / 1152.0
    u3 = (30375 - 369603*p2 + 765765*p4 - 425425*p6) * pp * p2 / 414720.0
    u4 = (4465125 - 94121676*p2 + 349922430*p4 - 446185740*p6 + 185910725*p8) * p4 / 39813120.0
    u5 = (1519035525.0 - 49286948607.0*p2 + 284499769554.0*p4   - 614135872350.0*p6 + 566098157625.0*p8 - 188699385875.0*p10) * pp * p4 / 6688604160.0
    u_corr_i = 1.0 + u1/v + u2/v2 + u3/v3 + u4/v4 + u5/v5

    # -z for ive
    result = log_res + np.log(u_corr_i) - abs(z)
    return result


for i, (v, z) in enumerate([
    (3, 1), (3, 5), (3, 10),
    (3, 50), (3, 100), (3, 200),
    (10, 1), (10, 5), (10, 10),
    (50, 1), (50, 5), (50, 10),
    (150, 1), (150, 5), (150, 10),
]):
    with np.errstate(all='ignore'):
        ive = scipy.special.ive(v, z)
        log_ive = np.log(ive)
        log_ive2 = test_ive_asy(v, z)
        err = (log_ive2 - log_ive) / log_ive
        print(f"  ive({v:3}, {z:2}) = {ive:9.2e}, log = {log_ive:.2f} vs {log_ive2:.2f} err: {err:.1e}")
        if i % 3 == 2: print()

```
</details>

<details>
<summary>Click to toggle results</summary>

```
  ive(  3,  1) =  8.16e-03, log = -4.81 vs -4.81 err: 4.5e-06
  ive(  3,  5) =  6.96e-02, log = -2.66 vs -2.66 err: 3.5e-06
  ive(  3, 10) =  7.98e-02, log = -2.53 vs -2.53 err: -2.0e-07

  ive(  3, 50) =  5.16e-02, log = -2.96 vs -2.96 err: 1.1e-11
  ive(  3, 100) =  3.82e-02, log = -3.27 vs -3.27 err: 1.7e-13
  ive(  3, 200) =  2.76e-02, log = -3.59 vs -3.59 err: 7.1e-15

  ive( 10,  1) =  1.01e-10, log = -23.01 vs -23.01 err: -4.9e-11
  ive( 10,  5) =  3.09e-05, log = -10.39 vs -10.39 err: -6.6e-10
  ive( 10, 10) =  9.94e-04, log = -6.91 vs -6.91 err: 8.2e-10

  ive( 50,  1) =  1.08e-80, log = -184.13 vs -184.13 err: -0.0e+00
  ive( 50,  5) =  1.98e-47, log = -107.54 vs -107.54 err: -5.3e-16
  ive( 50, 10) =  2.16e-34, log = -77.52 vs -77.52 err: 3.8e-15

  ive(150,  1) =  0.00e+00, log = -inf vs -709.99 err: nan
  ive(150,  5) = 6.03e-206, log = -472.54 vs -472.54 err: -0.0e+00
  ive(150, 10) = 6.57e-163, log = -373.44 vs -373.44 err: 1.5e-16
```
</details>

### PR DESCRIPTION
**Pull request by sethtroisi**
Fixes #12607 

I'm not sure if this is the right structure, the right place, or the right name.

1. `logiv`, `log_iv`, `lniv`, `ivln`
context: [scipy.special](https://docs.scipy.org/doc/scipy/reference/special.html) contains `ndtr(x)` and `log_ndtr(x)` | `gamma(z)`, `gammaln(x)`, and `loggamma(z)` | `beta(a, b)`, `betaln(a, b)`

1. I'm not sure what precision goal I should aim for;
    1. percent error seem to be very small when `v` is large but can be as high as `1e-9` when v is just greater than the threshold.

1. I'm not sure if I need to support complex numbers for these functions.

1. I don't yet support array/arraylike input.
    1. Is there a guide or example PR someone can point me at?

### CODE REVIEW

**Comment by sethtroisi** [comment]
@WarrenWeckesser I'd like to hear your opinion if these are worth including.

**Comment by sethtroisi** [comment]
a number of the windows targets fail due to `M_PI` being undeclared which is potentially related related to cython version 0.29.18 (see https://github.com/scikit-image/scikit-image/issues/4727)?

However _digamma.pxd uses`M_PI` and I doesn't fail. I looked in setup.py and I don't see any mention or special handling of _digamma

File: scipy/special/log_iv.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+from scipy.special import iv, ive
+from ._logiv import log_iv_asym
+
+__all__ = ['logiv', 'logive']
+
+def logiv(v, z):
+    """
+    Log of modified Bessel function of the first kind of real order.
+
+    See `iv` for Parameters and return
```
Comment (ev-br) [inline scipy/special/log_iv.py:24]: Provably want to add a See Also section

File: scipy/special/log_iv.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#
+# Author:  Seth Troisi 2020
+
+from numpy import log
+from scipy.special import iv, ive
+from ._logiv import log_iv_asym
+
+__all__ = ['logiv', 'logive']
+
+def logiv(v, z):
```
Comment (ev-br) [inline scipy/special/log_iv.py:20]: Maybe move the dispatch to Cython, so that cython_special interface is consistent.

**Review by ev-br — COMMENTED** [review]
A couple of drive-by comments, not a definitive review:

- it'd be nice to deduplicate with _hyp0f1, if possible

- the test_mpmath machinery is very useful for smoking out corner cases. Would be nice to include this

- best ensure that what you add here is available from the cython_special interface.

- these functions should be ufuncs, so cython kernels need to be hooked up into the ufunc-generation machinery

**Comment by sethtroisi** [comment]
> A couple of drive-by comments, not a definitive review:

@ev-br These are very useful.

> 
> * it'd be nice to deduplicate with _hyp0f1, if possible

I keep this in mind but won't look at it till we're closer to the end.

> * the test_mpmath machinery is very useful for smoking out corner cases. Would be nice to include this

I wasn't aware of this test; I added a `test_logiv` method which confirms accurate to `3e-14`

> * best ensure that what you add here is available from the cython_special interface.
> * these functions should be ufuncs, so cython kernels need to be hooked up into the ufunc-generation machinery

Can you point me at some links for doing this? I'm trying to follow https://github.com/scipy/scipy/commit/4e8635247518d9ca14340cbc33d1338b98ed8066 but I'm still missing something.

**Comment by ev-br** [comment]
> I keep this in mind but won't look at it till we're closer to the end.

Makes sense!

Re ufunc: you'll need to add an entry to functions.json, the docstrings go to add_newdocs.py. 

I'm not sure about the moving parts now: It used to be necessary to run `generate_ufuncs.py` manually, not sure if it's still needed (if it is, please make a separate GEN commit). 

The most authoritative review should be from Josh or Pauli.

**Comment by sethtroisi** [comment]
> Re ufuncs: you'll need to add an entry to functions.json, the docstrings go to add_newdocs.py.
> 
> I'm not sure about the moving parts now: It used to be necessary to run `generate_ufuncs.py` manually, not sure if it's still needed (if it is, please make a separate GEN commit).

I don't believe this is needed anymore; `_generate_pyx.py` seems to be run during `setup.py` 

I defined my function in `_logiv.pxd` as `cdef double log_iv_asym(double v, double z) nogil:` and added that definition (`dd->d`) to `functions.json`. I can import my function as `from ._ufuncs import _log_iv_asym` but I'm not able to pass an ndarray as parameter (I suspect and I'm missing a ufuncs wrapper function somewhere)

**Comment by ev-br** [comment]
Seems to work?

```
In [8]: from scipy.special._ufuncs import _log_iv_asym                          

In [9]: _log_iv_asym([1, 2, 3], [2, 3, 5])                                      
Out[9]: array([0.46121925, 0.80862499, 2.33515425])
```

However, I'm not sure what is the end goal here: is it really useful to add the asymptotic expansion ufunc to the API (albeit a private part of it). I suspect you'll change it later so that `from scipy.special import log_ive` works without a python wrapper.

**Comment by fancidev** [comment]
@dschmitz89 The log BesselI function would be very useful. I also wanted it (as well as log BesselK function) when I worked on a compound Poisson process. The `ive` and `kve` functions were helpful but they still overflow / underflow for large $x$ if I remember correctly.

However, I guess it’s quite challenging to write a solid `logiv`. I’ll probably do some research after I finish some of the other easier ones, especially the univariate “all roots” finder which I’m quite interested in.

**Comment by ehou-str** [comment]
Anyone know when this will be merged in? I agree it would be super useful for Von Mises-Fisher distributions in high dimensions when the ive function underflows. Currently scipy.stats.vonmises_fisher._log_norm_factor returns NaN for high order values.

**Comment by dschmitz89** [comment]
@ehou-str : since I implemented the von Mises Fisher, I am curious what your use case is for such a high number of dimensions.

Realistically, this PR will not land anymore, it has been inactive for a few years.

**Comment by ehou-str** [comment]
> @ehou-str : since I implemented the von Mises Fisher, I am curious what your use case is for such a high number of dimensions.
> 
> Realistically, this PR will not land anymore, it has been inactive for a few years.

I am looking at using the distribution with embedding spaces of Large Language Models (which vary from 700-ish to 4000+) where the embeddings are normalized for cosine similarity and thus lie on the surface of a hypersphere. 

The scipy.stats.vonmises_fisher._log_norm_factor which uses ive fails at around 500 dimensions and returns 0, which then gives NaN after the log. 

I ended up reimplementing some of the von Mises Fisher functions with a modified log_norm_factor function that uses a finite approximation to the summation based on the pyro implementation of log iv here: https://docs.pyro.ai/en/stable/_modules/pyro/ops/special.html#log_I1

**Comment by dschmitz89** [comment]
> > @ehou-str : since I implemented the von Mises Fisher, I am curious what your use case is for such a high number of dimensions.
> > Realistically, this PR will not land anymore, it has been inactive for a few years.
> 
> I am looking at using the distribution with embedding spaces of Large Language Models (which vary from 700-ish to 4000+) where the embeddings are normalized for cosine similarity and thus lie on the surface of a hypersphere.
> 
> The scipy.stats.vonmises_fisher._log_norm_factor which uses ive fails at around 500 dimensions and returns 0, which then gives NaN after the log.
> 
> I ended up reimplementing some of the von Mises Fisher functions with a modified log_norm_factor function that uses a finite approximation to the summation based on the pyro implementation of log iv here: https://docs.pyro.ai/en/stable/_modules/pyro/ops/special.html#log_I1

Interesting, had not expected the distribution to be used for such purposes!

Would you by any chance be interested to contribute this approximation to an internal Python helper function for `log(ive)`? We could dispatch to it for `v>500` and `k>1e9` where `ive` fails currently. Implementing it in `special` would be a bigger undertaking.

**Comment by fbourgey** [comment]
I am closing this. A new logiv/logive implementation should now start in https://github.com/scipy/xsf
